### DISCUSSION TITLE
Update profile information form with keycloak-js adapter

### DISCUSSION THREAD

**Discussion by omar-zahid** [keycloakify/keycloakify#608] [General]
According to the video on page https://docs.keycloakify.dev/account-theme, it's possible to access the update profile page using keycloak-js adapter. However, I can't find any method that redirects to this page.
For technical reasons, we can't use OIDC SPA library just yet.

![image](https://github.com/user-attachments/assets/8941bdb9-bfcf-45a9-a254-1a59251bd39d)

**Comment by garronej** [comment]
Hello @omar-zahid,  

It should look something like this (assuming chatGPT does not allucinate):  

```js
const keycloak = new Keycloak({
  url: 'https://your-keycloak-server/auth',
  realm: 'your-realm',
  clientId: 'your-client-id',
});

keycloak.init({
  onLoad: 'check-sso', // or 'login-required'
  checkLoginIframe: false
}).success(authenticated => {
  if (!authenticated) {
    keycloak.login({
      locale: 'fr', // If your Keycloak server is configured to recognize this parameter
      redirectUri: 'https://your-app-url', // Your redirect URL
      extraParams: {
        kc_action: "UPDATE_PROFILE"
      }
    });
  }
}).error(() => {
  console.log('Failed to initialize Keycloak');
});
```  

Just out of curiosity, what's the technical limitation that prevent you from migrating to oidc-spa?

**Reply by omar-zahid** [reply]
Thanks @garronej for pointing out that the login method accepts an action.

Here's the solution:

login method accepts `KeycloakLoginOptions`
```typescript
  /**
   * If value is `'register'` then user is redirected to registration page,
   * otherwise to login page.
   */
  action?: string;
```

Usage: Call the login method as per following (I have it exported from a user service library):

```typescript
function updateUserProfile() {
  return UserService.login({
    action: "UPDATE_PROFILE"
  });
}
```

Technical limitation that prevents us from migrating to oidc-spa is just that we've created a shared user service library that have most commonly used functionality exported and its being used by 4 to 5 apps at the moment in a monorepo. Migrating will take quite some effort. A bit disappointed that Keycloak will deprecate this library soon.

**Reply by garronej** [reply]
Thanks for sharing your solution, chatGPT did hallucinate, like it often does.  

> Technical limitation that prevents us from migrating to oidc-spa is just that we've created a shared user service library that have most commonly used functionality exported and its being used by 4 to 5 apps at the moment in a monorepo. Migrating will take quite some effort. A bit disappointed that Keycloak will deprecate this library soon.  

I understand your concerns. Don't worry too much about the deprecation. Although they do plan to eventually deprecate Keycloak-js and recommend a community-maintained alternative (hopefully oidc-spa), Keycloak-js will continue to work. While it won't receive any new updates, it is essentially stable software, so there's no urgent need to migrate.

The main motivation for migrating to oidc-spa should be its improved token refresh management and better handling of edge cases, leading to a more reliable user experience. However, if you’ve got Keycloak-js working well for your needs, there’s no need to rush into a migration. If it isn't broken, there's no need to fix it.
